src/actions/measurements.ts

```diff
@@ -611,14 +612,23 @@ export const applyMeasurementsColorBy = (
     }
   }
 
-  if (controls.measurementsColorGrouping !== undefined) {
-    dispatch({type: REMOVE_METADATA, nodeAttrsToRemove: [encodeMeasurementColorBy(controls.measurementsColorGrouping)]});
-  }
-  if (controls.measurementsColorGrouping !== groupingValue) {
-    dispatch({type: CHANGE_MEASUREMENTS_COLOR_GROUPING, controls:{measurementsColorGrouping: groupingValue}});
-  }
-  dispatch({type: ADD_EXTRA_METADATA, newNodeAttrs, newColorings});
-  dispatch(changeColorBy(measurementColorBy));
+  /**
+   * Batching all dispatch actions together to prevent multiple renders
+   * This is also _required_ to prevent error in calcColorScale during extra renders:
+   * 1. REMOVE_METADATA removes current measurements coloring from metadata.colorings
+   * 2. This triggers the componentDidUpdate in controls/color-by, which dispatches changeColorBy.
+   * 3. calcColorScale throws error because the current coloring is no longer valid as it was removed by REMOVE_METADATA in step 1.
+   */
+  batch(() => {
```

Comment:
I don't think it's worth changing what you've done (which includes a good docstring) but an alternative approach to consider for patterns like this if we use them again is to make "remove metadata" a thunk which checks that the attr being removed isn't currently used as a coloring, and if it is then thee thunk can do something about it.